include/mbedtls/ssl.h

```diff
@@ -4923,10 +4923,11 @@ int mbedtls_ssl_get_session(const mbedtls_ssl_context *ssl,
  *                 currently being processed might or might not contain further
  *                 DTLS records.
  *
- * \note           If the context is configured to allow TLS 1.3, or if
- *                 #MBEDTLS_USE_PSA_CRYPTO is enabled, the PSA crypto
+ * \note           If #MBEDTLS_USE_PSA_CRYPTO is enabled, the PSA crypto
  *                 subsystem must have been initialized by calling
  *                 psa_crypto_init() before calling this function.
+ *                 Otherwise, the handshake may call psa_crypto_init()
+ *                 if it ends up negotiating TLS 1.3.
```

Comment:
>   Otherwise, the handshake may call psa_crypto_init() if it ends up negotiating TLS 1.3.

Actually that's not quite accurate: the handshake may call `psa_crypto_init` if it _tries_ to negotiate TLS 1.3. I left it more open in `mbedtls_config.h`.

> So, wouldn't it be simpler to just base the call on whether 1.3 is enabled on our side or not 

Since this is an update to an LTS, I wanted to minimize behavior changes. It's impossible to only add crypto init calls to workflows that do negotiate TLS 1.3, because the negotiation might involve crypto. The most obvious example is that a client needs PSA crypto to send a 1.3 ClientHello, and the server might be 1.2-only. But I wanted to avoid changes that weren't necessary. In particular, a server getting a 1.2-only ClientHello won't start calling `psa_crypto_init`.

Reply:
> putting a single call neat the top of mbedtls_ssl_handshake_step()

I wanted to only change 1.3-specific files, so it would be obvious that builds without 1.3 weren't affected. Sure, an `#if defined(MBEDTLS_SSL_PROTO_TLS1_3)` also makes it locally obvious, but that means yet another ifdef so the code is more complex.

Reply:
Can we be more precise and say:

* A client calls `psa_crypto_init` if it sends a 1.3 hello;
* A server calls `psa_crypto_init` if it has 1.3 enabled and receives a 1.3 hello?

I wasn't sure whether we could reliably commit to that. I'm not even 100% sure that this is what's currently happening in all cases.

Reply:
> Can we be more precise

I'm actually inclined to be less precise and give ourselves more freedom to change things in the future if needed or just if we want to simplify the code.

I think "may call psa_crypto_init()` if TLS 1.3 is enabled" is good enough. I don't think a user needs to know more than that.

Reply:
> > putting a single call neat the top of mbedtls_ssl_handshake_step()
> 
> I wanted to only change 1.3-specific files, so it would be obvious that builds without 1.3 weren't affected.

Yes, and that's indeed one thing I like about your solution. Mostly I wanted to make the thinking explicit about why they're placed here.

Reply:
Ah, so we do. In an earlier attempt I had the initialization later, but it was too late for when the client sends a key share in ClientHello for PSK or session resumption.

I think it would be nice if a server operating in a 1.2-only environment would not change its behavior, but that would be more complicated to arrange. I think we'd have to have multiple places calling crypto_init in the server code.

Reply:
>I think it would be nice if a server operating in a 1.2-only environment would not change its behavior, but that would be more complicated to arrange. I think we'd have to have multiple places calling crypto_init in the server code.

I agree it would be nice, especially in an LTS where we're supposed to minimize changes. I was also afraid that would mean having to call `crypto_init` in more places, but after taking a look, I think there's an ideal place in `ssl_tls13_parse_client_hello()`: [just when we've determined this is a 1.3 Hello](https://github.com/Mbed-TLS/mbedtls/blob/development/library/ssl_tls13_server.c#L1409) - before that the function only does minimal parsing (memorising pointers and lengths to interesting parts of the message to come back to later when we know the negotiated version).

Wdyt?

Reply:
Thanks. I'll try that, with the current code as a fallback.

Reply:
> Can we be more precise and say:
> A client calls `psa_crypto_init` if it sends a 1.3 hello;
> A server calls `psa_crypto_init` if it has 1.3 enabled and receives a 1.3 hello?

In short I'd say, yes let's definitely try to do that, but no let's not announce it in the documentation - I don't think we'll have time to do enough testing to ensure we succeeded, and I'm rather not tie our hands if we want/need to change things in the future.